### PR TITLE
add full line comments

### PR DESCRIPTION
**Pull request by robartsd**
Blank lines and comment only lines are no longer flagged as errors in the routine debugger.

### CODE REVIEW

**Comment by ArnoldSmith86** [comment]
PR-SERVER-BOT: You can play around with it here: http://212.47.248.129:1089/pr-test (or any other room on that server)

**Review by rogerl50 — APPROVED** [review]
This is actually preferable to the Note hack. Note that if you want a multiple-line comment to sort the way it was written, you'll have to do something like
```JSON
  "//1 First part"
  "//2 Second part"
```
Not a complaint, just a note. 

Code look good as well.

**Comment by ArnoldSmith86** [comment]
> Note that if you want a multiple-line comment to sort the way it was written, you'll have to do something like [...]

If that was the case, all operations in a routine would have random order.

**Comment by rogerl50** [comment]
> > Note that if you want a multiple-line comment to sort the way it was written, you'll have to do something like [...]
> 
> If that was the case, all operations in a routine would have random order.

Good point.

**Comment by robartsd** [comment]
It would be nice if the debugger showed the full comment rather than just the part before the first space.

**Comment by rogerl50** [comment]
I'll take a look at that in the morning.

File: client/js/widgets/widget.js [hunk 1]
```diff
@@ -672,7 +672,8 @@ export class Widget extends StateManaged {
             variables[variable] = getValue(variables[variable]);
           if(jeRoutineLogging) jeLoggingRoutineOperationSummary(a.substr(4), JSON.stringify(variables[variable]));
         } else if(a.match(new RegExp('^ *(?://.*)?\x24'))) {
-          // ignore blank and comment only lines
+          // ignore (but log) blank and comment only lines
+          if(jeRoutineLogging) jeLoggingRoutineOperationSummary(a.substr(3));
```
Comment (robartsd) [inline client/js/widgets/widget.js:676]: This is assuming that the comment starts with a space.  `"//comment"` in a routine is logged as "//comment omment".

**Comment by rogerl50** [comment]
Modifed both `widgets.js` and `jsonedit.js` to properly handle comments and errors. Previous versions did not deal properly with comments with leading spaces, or with comments with no space following `//`. There may be a better way to do this (particularly in jsonedit.js); more review would be useful.

File: client/js/widgets/widget.js [hunk 2]
```diff
@@ -603,7 +603,9 @@ export class Widget extends StateManaged {
     for(const original of routine) {
       let a = JSON.parse(JSON.stringify(original));
       if(typeof a == 'object')
-        a = evaluateVariablesRecursively(a);
+        a = evaluateVariablesRecursively(a)
+      else
+        a = original.trim();
```
Comment (robartsd) [inline client/js/widgets/widget.js:608]: This does mean that leading spaces before a variable expression are now allowed where they previously were not. (I don't care, but someone might.)

File: client/js/widgets/widget.js [hunk 3]
```diff
@@ -672,7 +674,13 @@ export class Widget extends StateManaged {
             variables[variable] = getValue(variables[variable]);
           if(jeRoutineLogging) jeLoggingRoutineOperationSummary(a.substr(4), JSON.stringify(variables[variable]));
         } else {
-          problems.push('String could not be interpreted as expression. Please check your syntax and note that many characters have to be escaped.');
+          const comment = a.match(new RegExp('^ *(?://(.*))?\x24'));
```
Comment (robartsd) [inline client/js/widgets/widget.js:677]: Since trim is used above, the RegExp does not need the leading ` *` to match leading spaces.

**Comment by robartsd** [comment]
> Modifed both `widgets.js` and `jsonedit.js` to properly handle comments and errors. Previous versions did not deal properly with comments with leading spaces, or with comments with no space following `//`. There may be a better way to do this (particularly in jsonedit.js); more review would be useful.

I do like the more consistent formatting in the debugger.

**Review by 96LawDawg — APPROVED** [review]
Approved through test.

File: client/js/widgets/widget.js (hunk 2 above)
Comment (rogerl50) [reply client/js/widgets/widget.js:608]: Yes. I meant to mention this in my comment but forgot. I actually think it's a feature, not a bug.

File: client/js/widgets/widget.js (hunk 3 above)
Comment (rogerl50) [reply client/js/widgets/widget.js]: Fixed.

File: client/js/jsonedit.js [hunk 4]
```diff
@@ -1422,7 +1422,17 @@ function jeLoggingRoutineEnd(variables, collections) {
...
+  if (typeof applied == 'string')
+    if (applied.substring(0,3) == 'var')
+      fcn = 'var'
+    else if (applied.substring(0,2) == '//')
+      fcn = '//'
+    else
+      fcn = applied
+  else
+    fcn = applied.func || '<COMMENT>'
+  jeHTMLStack.unshift([jeLoggingHTML, original, applied, fcn]);
```
Comment (ArnoldSmith86) [inline client/js/jsonedit.js:1435]: Looks like you removes the `html()` call around it which makes sure that HTML tags in the comment are escaped.

Comment (rogerl50) [reply client/js/jsonedit.js]: Oops.

**Pull request merged**